astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -179,6 +179,35 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> Sequence[Us
         return user_defined_functions
 
 
+@dataclasses.dataclass
+class KeeperMapTablesReadOnlyStep(Step[None]):
+    clients: Sequence[ClickHouseClient]
+    allow_writes: bool
+
+    @staticmethod
+    def get_revoke_statement(table: Table, escaped_user_name: str) -> bytes:
+        return f"REVOKE INSERT, UPDATE, DELETE ON {table.escaped_sql_identifier} FROM {escaped_user_name}".encode()
+
+    @staticmethod
+    def get_grant_statement(table: Table, escaped_user_name: str) -> bytes:
+        return f"GRANT INSERT, UPDATE, DELETE ON {table.escaped_sql_identifier} TO {escaped_user_name}".encode()
+
+    async def run_step(self, cluster: Cluster, context: StepsContext):
+        _, tables = context.get_result(RetrieveDatabasesAndTablesStep)
+        replicated_users_response = await self.clients[0].execute(
+            b"SELECT base64Encode(name) FROM system.users WHERE storage = 'replicated' ORDER BY name"
+        )
+        replicated_users_names = [b64decode(cast(str, user[0])) for user in replicated_users_response]
+        keeper_map_table_names = [table for table in tables if table.engine == "KeeperMap"]
+        privilege_altering_fun = self.get_grant_statement if self.allow_writes else self.get_revoke_statement
+        statements = [
+            privilege_altering_fun(table, escape_sql_identifier(user))
+            for table in keeper_map_table_names
+            for user in replicated_users_names
```

Comment:
1. this looks a bit dangerous, because num_users*num_tables may blow up for some cases which we do not control.
2. and it might be not only just long, but will grow replication queue very fast
3. This replication queue not only should be stored, but also replayed by new replicas, repeating all those steps (as I understand)

Reply:
I don't think that's a problem in practice, we usually have few users and little to no KeeperMap tables.